src/combinatorics/ramsey/infinite.lean

```diff
@@ -0,0 +1,468 @@
+
+/--
+There exists an infinite sequence of homogeneous projections,
+each a refinement of the previous one.
+-/
+lemma exists_homo_proj_seq (f : [ℕ]² → color) :
+∃ (g : ℕ → homogeneous_proj f), init f < g 0 ∧ ∀ n, g n < g (n+1) :=
+begin
+  have ac : ∃ (cf : Π (x : homogeneous_proj f), (λ (x : homogeneous_proj f), homogeneous_proj f) x),
+    ∀ (p : homogeneous_proj f), (λ (p q : homogeneous_proj f), p<q) p (cf p),
+    exact axiom_of_choice (refine_homo_proj f),
+  cases ac with cf Hcf,
+  let g := λ m : ℕ, (iterate_refinement f cf m),
+  apply exists.intro g,
+  constructor,
+  { exact Hcf (init f) },
+  { intro n,
+    have h1 : (g n) < cf (g n),
+      exact Hcf (g n),
+    have h2 : ∀ p : homogeneous_proj f, p = ⟨f, ↑p, p.pt⟩,
+      intro p, cases p with H n, cases H with s pf, refl,
+    have h3 : cf (g n) = iterate_refinement f cf (n+1),
+      unfold iterate_refinement, simp, rw ← (h2 (g n)),
+    simp * at * }
+end
+
+/--
+The sets in a sequence of refined homogeneous projections are
+closed upward under inclusion.
+-/
+lemma homo_proj_seq_mono_sets
+  (f : [ℕ]² → color)
+  (g : ℕ → homogeneous_proj f)
+  (h : ∀ n, g n < g (n+1))
+  (x y : ℕ) :
+  (g (x+y)).s ⊆ (g x).s :=
+begin
+  induction y with y ih,
+  { intros y hy, exact hy },
+  { intros a ha,
+    exact ih ((h (x+y)).right.right.left ha) }
+end
+
+/--
+The points in a sequence of refined homogeneous projections are
+contained in the previous sets.
+-/
+lemma homo_proj_seq_mono_pts
+  (f : [ℕ]² → color)
+  (g : ℕ → homogeneous_proj f)
+  (h : ∀ n, g n < g (n+1))
+  (x y : ℕ) :
+  (g (x+y+1)).pt ∈ (g x).s :=
+(homo_proj_seq_mono_sets f g h x y) ((h (x+y)).right.left)
+
+/--
+The points in a sequence of refined homogeneous projections have
+the property that f {xᵢ, y} = f {xᵢ, z} for all xᵢ < y < z in Sₓ.
+-/
+lemma homo_proj_seq_stable_colors
+  (f : [ℕ]² → color)
+  (g : ℕ → homogeneous_proj f)
+  (h0 : (init f) < g 0)
+  (hn : ∀ n, g n < g (n+1))
+  (x y : ℕ) :
+  (project f (g x).pt) (g (x+1)).pt = (project f (g x).pt) (g (x+y+1)).pt :=
+begin
+  cases x,
+  { have h : homogeneous (project f (g 0).pt) (g 0),
+      exact h0.right.right.right,
+    cases h with c hm,
+    simp,
+    rw hm ((g 1).pt) (homo_proj_seq_mono_pts f g hn 0 0),
+    have hzy : y + 1 = 0 + y + 1, simp,
+    rw hzy,
+    rw hm ((g (0+y+1)).pt) (homo_proj_seq_mono_pts f g hn 0 y) },
+  { have hgx : g x < g (x+1), exact (hn x),
+    have hm : homogeneous (project f (g (x+1)).pt) (g (x+1)),
+      exact hgx.right.right.right,
+    cases hm with c Hm',
+    rw Hm' ((g (x + 1 + 1)).pt) (homo_proj_seq_mono_pts f g hn (x+1) 0),
+    rw Hm' ((g (x + 1 + y + 1)).pt) (homo_proj_seq_mono_pts f g hn (x+1) y) }
+end
+
+section increasing_functions
+
+--TODO are the results of this section proved elsewhere in mathlib?
+
+def increasing (f : ℕ → ℕ) := ∀ x y : ℕ, x < y → f x < f y
+
+lemma increasing_by_step (f : ℕ → ℕ) :
+  (∀ n : ℕ, f n < f (n+1)) → increasing f :=
```

Comment:
This is `strict_mono.nat` in `algebra.order_functions`. I see `incr_dom` exists as special case of `strict_mono.lt_iff_lt`, but the other ones are missing (but welcome in that file)

Reply:
thank you! `strict_mono` and `lt_iff_lt` are perfect. I think the other lemmas do make sense in `algebra.order_functions`, but I'll need to think about how to prove them with the right generalization, and without `nat.rec_on`.